### PR TITLE
[release/0.4][Performance] Split kv_b_proj and drop the strided copy before grouped matmul

### PR DESCRIPTION
**Pull request by DanielSun11**
Merged in dev: https://github.com/PaddlePaddle/PaddleFleet/pull/1696

### PR Category
Performance Optimization

### PR Types
Performance

### Description

latent MQA 路径上两处优化。

**1. 拆分 `kv_b_proj` 为独立吸收参数**

新增开关 `mqa_split_kv_b_proj`（仅 `hybrid_mla_attention` 为 `mqa_dsa` / `mqa_full_causal` 时有效）。开启后不再每次 forward 切片 `kv_b_proj.weight` 再走 `einsum`，改用两个已按 `fused_grouped_matmul` 布局存好的参数：`k_b_proj`（query 吸收）与 `v_b_proj`（V 去吸收），每侧一次 grouped GEMM，无切片、无转置。受 AOA 不能改 rank 限制，二者以折叠前两维的 2-D 存储，forward 内零拷贝 reshape 还原；`kv_b_proj` 在该模式下无梯度，缺这两个 key 的 checkpoint 由 loader（`_mla_split_kv_b_statements`）拆出。Muon 侧同步注册逐 head 切片（`axis=-2`），V 块转置回未拆分权重的朝向后再正交化，以保持各 `muon_version` 的缩放语义。

**2. `fused_grouped_matmul` 不再拷贝输入**

kernel 本身接收全部 stride，普通 strided view（如 `q[..., :qk_nope_head_dim]`）无需 `.contiguous()`。原实现每次 forward（含 recompute）及 backward 的 `dy` 都会白拷一次，而 phi 的 strided→dense kernel 无向量化特化（~350 GB/s vs dense ~4400 GB/s），代价与体积不成比例。新增 `_grouped_3d_strides` 只接受“dense 拷贝会逐元素线性化”的布局，broadcast / 负 stride / 自重叠 / 不可折叠一律回退原路径；入口另加 `w` 为 3-D `[G, R, D]`、`x.shape[-2:] == (G, D)` 的校验。

性能（`[1, 16384, 8, 128]` bf16）：fwd+bwd 1.080 → 0.869 ms/step（**1.24x**），`contiguous()` 调用 2/step → 0；8k profile 中约省 266 ms GPU 时间。

Cherry-pick of #1696 to `release/0.4`.

Merged in dev: #1696  

### 是否引起精度变化
否

零拷贝路径喂给同一 kernel 的值、顺序与 block 配置完全相同，结果**逐 bit 一致**：强制回退到拷贝路径后对比 `out` / 输入梯度 / `w.grad` 的 md5，312 种组合（13 种输入布局 × 2 dtype × 对齐与非对齐尺寸 × dense/非连续 `dy` × 三种冻结组合）与 3 个生产调用点全部零差异。拆分模式的等价性由 `TestSplitKvBProj` 覆盖：吸收 query、去吸收输出、整层 forward（含 warmup 的 eval/no-grad 早退路径）相对误差均 < 1e-3。

### CODE REVIEW

File: src/paddlefleet/transformer/mqa_latent_attention.py [hunk 1]
```diff
@@ -447,13 +480,25 @@ def _sparse_attn(self, query, kv, token_indices, sm_scale, d_v):
         )
 
     @staticmethod
-    def _deabsorb(core_out, v_b_proj_weight) -> Tensor:
+    @staticmethod
+    def _deabsorb(core_out, v_b_proj_weight, split_kv_b=False) -> Tensor:
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/mqa_latent_attention.py:831]: P1 该布局分支还有一个旧调用点未传 `split_kv_b`：`_forward_dsa` 的 warmup/eval 早退路径（当前第 552 行）仍调用 `_deabsorb(core_out, v_b_proj_weight)`。当 `dsa_indexer_use_sparse_loss=False` 且处于 eval、`no_grad` 或全层重计算的第一次 forward 时，新增的 `[h, v, l]` 权重会按旧 `[l, h, v]` 布局解包；生产尺寸 `h/v/l=64/256/512` 会把末维 `64*512` 的 `core_out` reshape 为 `256*64`，直接失败。请让该早退路径也传入 `self.split_kv_b`，并补一条覆盖 warmup 的 eval/no-grad（含 recompute 首次 forward）回归。

File: src/paddlefleet/triton_ops/grouped_matmul_fusion.py [hunk 2]
```diff
@@ -275,8 +314,15 @@ def forward(ctx, x, w):
         for s in orig_shape[:-2]:
             M *= s
 
-        # x is [..., G, D] -> [M, G, D], NO transpose needed
-        x_3d = x.reshape([M, G, D]).contiguous()
+        # x is [..., G, D] -> [M, G, D]. A foldable unit-stride view is read in
+        # place; anything else is materialized as a dense copy as before.
+        x_strides = _grouped_3d_strides(x)
```
Comment (risemeup1111) [inline src/paddlefleet/triton_ops/grouped_matmul_fusion.py:330]: P1 零拷贝分支绕过了原先 `reshape([M, G, D])` 隐含的输入/权重维度校验：这里的 helper 只按 `x` 自身的 shape/stride 判定，而后续 `G/D` 来自 `w`。例如 `x.shape=[1,3,6]`、stride 为 `(18,6,1)`、`w.shape=[3,R,4]` 时 helper 会返回 `(18,6,1)`，kernel 随后静默只读每组前 4 列；若权重的 `D` 更大还会跨组乃至越界读取，旧路径则会在 reshape 时拒绝。请在进入 helper/kernel 前显式校验 `x.shape[-2:] == (w.shape[0], w.shape[2])`（并校验权重 rank），并补不匹配维度的回归。

File: src/paddlefleet/transformer/multi_latent_attention.py [hunk 3]
```diff
@@ -1352,6 +1379,47 @@ def __init__(
...
+            # they cannot produce a 3-D tensor. Exposing a reshaped *view* from
+            # ``sharded_state_dict`` instead is not safe: DCP reshapes the
+            # ``ShardedWeight.local_tensor`` in place (``reshape_`` /``flatten_``
+            # at ``dcp/load_state_dict.py:557,741,998``), which on a view that
+            # aliases the parameter corrupts the parameter's own shape.
+            #
+            # The 3-D form is recovered per forward with a zero-copy reshape.
+            # TP is forced to 1 in this mode, so ``num_attention_heads_per_partition``
+            # is the full head count and the parameter needs no TP attributes.
+            self.k_b_proj = self.create_parameter(
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/multi_latent_attention.py:1443]: P1 这两个新参数都是 2-D，因此 Muon 会选择它们；但下方 `muon_slice_specs()` 仍只为旧的 `kv_b_proj.weight` 注册逐 head 切片，`k_b_proj` / `v_b_proj` 会落入 whole-matrix 更新。这样会把折叠在第 0 维的所有 head 当成一张矩阵正交化，破坏旧布局下每个 K/V head 独立更新的语义。请为直接参数键 `k_b_proj` 和 `v_b_proj` 注册按 `axis=-2`、`heads=num_heads` 的切片，并覆盖 Muon 单参数及同 shape 批处理后的 3-D 输入。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
| 序号 | 位置 | 优先级 | 状态 |
|---|---|---|---|
| 1 | [warmup/recompute 的 V 布局](https://github.com/PaddlePaddle/PaddleFleet/pull/1679#discussion_r3740751966) | P1 | ✅ |
| 2 | [grouped matmul 维度合约](https://github.com/PaddlePaddle/PaddleFleet/pull/1679#discussion_r3740752401) | P1 | ✅ |
| 3 | [split 参数的 Muon 切片](https://github.com/PaddlePaddle/PaddleFleet/pull/1679#discussion_r3740752866) | P1 | ✅ |
| 4 | [checkpoint 的 split 权重转换](https://github.com/PaddlePaddle/PaddleFleet/pull/1679#discussion_r3744596086) | P1 | 🚧 |

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/transformer/mqa_latent_attention.py (hunk 1 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/mqa_latent_attention.py:831]: 当前提交已在 `_forward_dsa` 的 warmup/eval 早退路径传入 `self.split_kv_b`，新增回归也覆盖了该调用参数；该问题已解决。

File: src/paddlefleet/transformer/multi_latent_attention.py (hunk 3 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/multi_latent_attention.py:1443]: 新增的 `axis=-2` 切片已避免跨 head 的 whole-matrix 更新，但 V 侧仍未保持全部受支持配置的旧语义：`v_b_proj` 的单 head 块是 `[v_head_dim, kv_lora_rank]`，旧权重块则是 `[kv_lora_rank, v_head_dim]`。当前实现也注明只有 `muon_version=3` 的缩放对转置对称；上层仍允许 1/2 并传入优化器，而这两版按 `dout/din` 缩放，转置后更新幅度会改变。请在 V 切片中转置到旧布局后调用 `ortho_fn` 再转回，或拒绝 split 模式与版本 1/2 的组合，并补对应回归。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: src/paddlefleet/transformer/mqa_latent_attention.py (hunk 1 above)
Comment (DanielSun11) [reply src/paddlefleet/transformer/mqa_latent_attention.py:831]: 确认已解决，补充一下当前状态：分支已与最新 `release/0.4` 合并，上游把该早退路径重构成了 `_forward_full_causal`（`mqa_latent_attention.py`），合并时我保留了本 PR 的语义，该函数的去吸收调用仍是 `self._deabsorb(core_out, v_b_proj_weight, self.split_kv_b)`；`_forward_sparse` 的调用点同样带上了 `self.split_kv_b`。warmup 的 eval/no-grad 回归 `TestSplitKvBProj::test_warmup_eval_forward_matches_unsplit` 保留并通过。

File: src/paddlefleet/triton_ops/grouped_matmul_fusion.py (hunk 2 above)
Comment (DanielSun11) [reply src/paddlefleet/triton_ops/grouped_matmul_fusion.py:330]: 意见成立，已修复。零拷贝分支确实丢掉了原先 `x.reshape([M, G, D])` 隐含的校验：`G`/`D` 来自 `w` 并作为 kernel 边界，而 `x` 只按自身 stride 读取，不匹配时会跨组甚至越界读，而不是报错。

修法是在 `fused_grouped_matmul` 入口显式校验（dense 路径和零拷贝路径都覆盖）：

- `w.ndim != 3` -> ValueError；
- `x.ndim < 2` 或 `tuple(x.shape[-2:]) != (w.shape[0], w.shape[2])` -> ValueError。

回归 `TestFusedGroupedMatmul::test_shape_mismatch_raises` 覆盖：错 `D`、错 `G`、非 2-D 的 `x`、2-D 的 `w`，外加一条正例——trailing 维正确的 strided view（`wide[..., :D]`）仍然走零拷贝路径并返回正确 shape，确认校验没有把合法的 view 一起拒掉。

本地提交 `d2b0633f`，随下一次 push 到本 PR。

File: src/paddlefleet/transformer/multi_latent_attention.py (hunk 3 above)
Comment (DanielSun11) [reply src/paddlefleet/transformer/multi_latent_attention.py:1443]: 意见成立，已按第一种方案修复：在 V 侧切片里转置到旧布局后再调用 `ortho_fn`，然后转回，而不是限制 `muon_version`。

核对了一下依据（`paddle/optimizer/muon.py`）：`_zeropower_via_newtonschulz5` 对 `shape[-2] > shape[-1]` 的块会先转置、迭代后再转回，所以 Newton-Schulz 本身对朝向不敏感；真正不对称的是 `_scaling_fn`——version 3 是 `max(dout, din) ** 0.5`（对称），version 1/2 分别是 `max(1, dout/din) ** 0.5` 和 `(dout/din) ** 0.5`，转置后就变成倒数比值。上层确实允许 1/2，所以不能只靠注释约定。

实现：`muon_utils` 新增 `_transposing` 包装，`ortho_blocks` / `ortho_per_head` 增加 `transposed` 参数（只交换最后两维，因此 Muon 批处理出的 3-D 输入也正确）；`v_b_proj` 的 spec 变为 `{"heads": num_heads, "axis": -2, "transposed": True}`，`k_b_proj` 保持不变（它本来就是旧布局的 K 块）。这样每个 V 头都在 `[kv_lora_rank, v_head_dim]` 朝向上被正交化并缩放，与未拆分权重逐 head 的更新语义一致，且对三个版本都成立。

回归：`test_hybrid_mla_grad_health.py::TestSplitKvBProj::test_muon_slices_the_split_params_per_head` 断言两个参数的 kwargs 与实际交给 `ortho_fn` 的每头块形状（`k_b_proj` -> `[kv_lora_rank, qk_nope_head_dim]`，`v_b_proj` -> `[kv_lora_rank, v_head_dim]`），并覆盖单参数与 stack 成 3-D 的批处理输入；`test_muon_slice_specs.py` 另加 `test_split_kv_b_specs_handle_muon_batched_3d_input`。

本地提交 `d2b0633f` / `0f432ad3`，随下一次 push 到本 PR。

File: src/paddlefleet/transformer/transformer_config.py [hunk 4]
```diff
@@ -866,6 +866,32 @@ class TransformerConfig(ModelParallelConfig):
...
+    *stored* with the leading two dims folded, i.e. as 2-D
+    ``[heads * kv_lora_rank, qk_nope_head_dim]`` and
+    ``[heads * v_head_dim, kv_lora_rank]``; the 3-D form is recovered per forward
+    with a zero-copy reshape. The fold exists because the AOA engine cannot
+    change a tensor's rank, so the checkpoint side has to be 2-D. Each side is
+    then one grouped Triton GEMM with no slice, no einsum and no transpose.
+    ``kv_b_proj`` receives no gradient at all in this mode and the parameter set
+    is no longer byte-compatible with a dense MHA phase: a checkpoint without
+    these two keys must have them split out of ``kv_b_proj.weight`` by the loader
+    (``_mla_split_kv_b_statements`` generates the AOA statements that do it).
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/transformer_config.py]: P1 当前开关还无法从现有 checkpoint 启动：生产 `fleet_model/ernie5_v2/modeling.py::_gen_aoa_config` 仍只把旧权重映射到目标 `kv_b_proj.weight`，当前仓库和该加载器中也都没有这里引用的 `_mla_split_kv_b_statements`。因此目标模型新增的 `k_b_proj` / `v_b_proj` 没有来源，AOA 在检查 destination keys 时会以 “needs to be loaded, but not found in checkpoint” 终止。现有 `TestSplitKvBProj` 通过测试内手写的 `_loader_split` 注入两项，未覆盖真实加载路径。请在实际 `_gen_aoa_config` 路径落地拆分 statements（或明确关联并先落地对应依赖），并补一条从仅含旧 `kv_b_proj.weight` 的 checkpoint 生成这两个目标张量的回归。

File: src/paddlefleet/transformer/mqa_latent_attention.py (hunk 1 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/mqa_latent_attention.py:831]: 已在当前 head 复核：`_forward_full_causal` 与 `_forward_sparse` 均传入 `self.split_kv_b`，warmup eval/no-grad 回归也保留；该问题已解决。

File: src/paddlefleet/triton_ops/grouped_matmul_fusion.py (hunk 2 above)
Comment (risemeup1111) [reply src/paddlefleet/triton_ops/grouped_matmul_fusion.py:330]: 已在当前 head 复核：入口同时校验 `w.ndim == 3` 与 `x.shape[-2:] == (G, D)`，回归覆盖错误 G/D、低 rank 输入及合法 strided view；该问题已解决。

File: src/paddlefleet/transformer/multi_latent_attention.py (hunk 3 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/multi_latent_attention.py:1443]: 已在当前 head 复核：V 侧通过 `transposed=True` 仅交换最后两维后调用 `ortho_fn`，并覆盖 2-D 单参数及 3-D 批处理输入；该问题已解决。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: src/paddlefleet/transformer/transformer_config.py (hunk 4 above)
Comment (DanielSun11) [reply src/paddlefleet/transformer/transformer_config.py]: erniebot还未合入

Comment (risemeup1111) [reply src/paddlefleet/transformer/transformer_config.py]: 了解。由于该 PR 已公开 `mqa_split_kv_b_proj`，且描述承诺旧 checkpoint 可由 loader 拆分，但对应 erniebot 改动尚未合入，当前可交付组合仍无法完成这条加载路径，因此此项先保持未解决。请在 erniebot 的提交/PR 可见后关联到这里，并提供实际 `_gen_aoa_config` 到 AOA 加载的回归证据，届时再按完整依赖复核。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

**Review by LiYuRio — DISMISSED** [review]

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
checkpoint 的 split 权重转换仍未落地，现有 P1 保持未解决，详见[既有审查线程](https://github.com/PaddlePaddle/PaddleFleet/pull/1679#discussion_r3744596086)。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
checkpoint 的 split 权重转换仍未落地，现有 P1 保持未解决，详见[既有审查线程](https://github.com/PaddlePaddle/PaddleFleet/pull/1679#discussion_r3744596086)。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
checkpoint 的 split 权重转换仍未落地，现有 P1 保持未解决，详见[既有审查线程](https://github.com/PaddlePaddle/PaddleFleet/pull/1679#discussion_r3744596086)。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
checkpoint 的 split 权重转换仍未落地，现有 P1 保持未解决，详见[既有审查线程](https://github.com/PaddlePaddle/PaddleFleet/pull/1679#discussion_r3744596086)。

**Review by LiYuRio — APPROVED** [review]

**Review by From00 — APPROVED** [review]
LGTM
